### PR TITLE
Liveness check starts after container is ready

### LINKED ISSUE #27114: LivenessProbe should start after ReadinessProbe Succeeded if ReadinessProbe is specified
Our initial understanding is that liveness probe will start to check after readiness probe was succeeded but it turn out not to be like that.

We are testing with our system that has a long boot time, an approximation of boot time is between 1-3 minutes. We specify readiness probe with same url of liveness probe and specify initial delay of liveness probe to be 30 seconds, we found that the pod is killed by failing of liveness probe while readiness probe still in a failure.

So, why don't we specify initial delay of liveness probe to be more than 3 minutes? Well, it might be a case when readiness probe succeed at first minute and after that the pod will fail to do a job before liveness probe will start. So, it will affect running service.

Another point is why we don't only put readiness probe and leave out liveness probe? When readiness probe fail again it might take out of service as well so, it don't cause any affect for running service but it won't be restarted and we have to do manual restart them.

There are some concerns we thought about, here are list:
- What if readiness probe never succeed while starting a pod, can we specify a maximum number of check will be performed and if readiness probe fail more than that number, we may consider that pod should be killed.
- What if someone want to set readiness probe to put the pod in maintenance mode, the pod might be killed by above point if readiness probe fail more than that number or liveness probe will kill it. Maybe, we add one more state of readiness probe called `Maintenance` so that liveness probe should stop working when enter this state and no number of failing readiness probe should not be considered.

What do you think?

### PR DESCRIPTION
**Pull request by fraenkel**
Allow liveness to begin probing when there is no initial delay once the
container is ready. Any non-zero initial delay will behave as it did
previously which starts probing when the container starts.

fixes #27114

```release-note
Liveness probe begins after the container is marked ready when the initial delay is 0 seconds.
```

### CODE REVIEW

**Comment by fraenkel** [comment]
@k8s-bot gce etcd3 e2e test this
@k8s-bot verify test this
@k8s-bot cvm gce e2e test this

File: pkg/kubelet/prober/worker.go [hunk 1]
```diff
@@ -56,6 +56,8 @@ type worker struct {
 	resultsManager results.Manager
 	probeManager   *manager
 
+	dependentManager results.Manager
```
Comment (mtaufen) [inline pkg/kubelet/prober/worker.go]: Why not call this `readinessManager`, if that's what it's used for?

Comment (fraenkel) [inline pkg/kubelet/prober/worker.go]: Only because there was no guarantee you passed in a readiness manager.  You could have others.
I will rename it.

**Review by yujuhong — CHANGES_REQUESTED** [review]
This PR would change the meaning of the readiness and liveness checks. A container can be alive but not ready, or not alive and not ready. With this PR, a container may never be killed and restarted if it's not ready. I don't think this should be allowed.

**Comment by fraenkel** [comment]
There is only 1 case where the behavior does change, otherwise it is exactly the same.
If you have readiness set and a liveness delay of 0, you get the waiting behavior. One can argue that is configuration doesn't even make sense in the current world since liveness begins regardless of readiness.

**Comment by yujuhong** [comment]
IIUC, liveness was intended to start before readiness in many cases. A process is running doesn't mean it's ready to serve traffic. It won't be restarted (since it passes liveness checks), but will not be listed as endpoints in a service.

**Comment by fraenkel** [comment]
I guess it all depends on how you do liveness. Many people were using probes so you need the process to be ready.
I was trying to avoid adding yet another flag to allow this behavior. I assume we should discuss how one can have liveness begin after readiness when that makes senses.

**Comment by yujuhong** [comment]
@fraenkel, I think we all agree that changing the semantics liveness/readiness checks may cause undesirable surprises to existing users. If you don't mind, let's close this PR and continue discussing on the issue to find a better solution.

**Comment by fraenkel** [comment]
Closing for further discussion.

**Pull request closed without merging**